File: src/test/java/at/bitfire/dav4android/exception/HttpExceptionTest.java
/*
 * Copyright © Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */

package at.bitfire.dav4android.exception;

import org.junit.Test;

import okhttp3.Protocol;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertTrue;

@SuppressWarnings("ThrowableInstanceNeverThrown")
public class HttpExceptionTest {

    final static String responseMessage = "Unknown error";

    @Test
    public void testHttpFormatting() {
        Request request = new Request.Builder()
                .post(RequestBody.create(null, "REQUEST\nBODY" + (char)5))
                .url("http://example.com")
                .build();

        Response response = new Response.Builder()
                .request(request)
                .protocol(Protocol.HTTP_2)
                .code(500)
                .message(responseMessage)
                .body(ResponseBody.create(null, (char)0x99 + "SERVER\r\nRESPONSE"))
                .build();
        HttpException e = new HttpException(response);
        assertTrue(e.getMessage().contains("500"));
        assertTrue(e.getMessage().contains(responseMessage));
        assertTrue(e.getRequest().contains("REQUEST[LF]\nBODY[05]"));
        assertTrue(e.getResponse().contains("[99]SERVER[CR][LF]\nRESPONSE"));
    }

}

File: src/test/java/at/bitfire/dav4android/exception/HttpExceptionTest.kt
/*
 * Copyright © Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */

package at.bitfire.dav4android.exception

import okhttp3.*
import org.junit.Assert.assertTrue
import org.junit.Test

class HttpExceptionTest {

    private val responseMessage = "Unknown error"

    @Test
    fun testHttpFormatting() {
        val request = Request.Builder()
                .post(RequestBody.create(null, "REQUEST\nBODY" + 5.toChar()))
                .url("http://example.com")
                .build()

        val response = Response.Builder()
                .request(request)
                .protocol(Protocol.HTTP_2)
                .code(500)
                .message(responseMessage)
                .body(ResponseBody.create(null, 0x99.toChar() + "SERVER\r\nRESPONSE"))
                .build()
        val e = HttpException(response)
        assertTrue(e.message!!.contains("500"))
        assertTrue(e.message!!.contains(responseMessage))
        assertTrue(e.request!!.contains("REQUEST\nBODY[05]"))
        assertTrue(e.response!!.contains("[99]SERVER↵\nRESPONSE"))
    }

}
